File: audit-imports.py
#!/usr/bin/env python3

import os
import re
import sys
from datetime import datetime
from pathlib import Path
from typing import List, Dict, Tuple

# Cores ANSI
class Colors:
    RED = '\033[31m'
    GREEN = '\033[32m'
    YELLOW = '\033[33m'
    BLUE = '\033[34m'
    MAGENTA = '\033[35m'
    CYAN = '\033[36m'
    RESET = '\033[0m'

# Configuração
CORRECT_IMPORT = '@/lib/db/models/'
EXCLUDE_DIRS = ['node_modules', '.next', 'dist', 'build', '.git']
EXTENSIONS = ['.ts', '.tsx']

# Padrões de import incorretos
WRONG_PATTERNS = [
    {
        'regex': re.compile(r"from\s+['\"]@/models/"),
        'description': "Import usando @/models/",
    },
    {
        'regex': re.compile(r"from\s+['\"]@/app/models/"),
        'description': "Import usando @/app/models/",
    },
    {
        'regex': re.compile(r"from\s+['\"]@/src/models/"),
        'description': "Import usando @/src/models/",
    },
    {
        'regex': re.compile(r"from\s+['\"]\.\/models/"),
        'description': "Import relativo ./models/",
    },
    {
        'regex': re.compile(r"from\s+['\"]\.\.\/models/"),
        'description': "Import relativo ../models/",
    },
    {
        'regex': re.compile(r"from\s+['\"]\.\.\/\.\.\/models/"),
        'description': "Import relativo ../../models/",
    },
]

# Estatísticas
stats = {
    'total_files': 0,
    'problematic_files': [],
    'issues_by_type': {},
    'total_issues': 0,
}

def should_ignore(path: str) -> bool:
    """Verificar se deve ignorar diretório/arquivo"""
    return any(excluded in path for excluded in EXCLUDE_DIRS) or 'backup_' in path

def scan_file(file_path: str) -> List[Dict]:
    """Escanear arquivo em busca de imports incorretos"""
    issues = []
    
    try:
        with open(file_path, 'r', encoding='utf-8') as f:
            lines = f.readlines()
            
        for line_num, line in enumerate(lines, 1):
            for pattern in WRONG_PATTERNS:
                if pattern['regex'].search(line):
                    # Sugerir correção
                    corrected = re.sub(
                        r"from\s+['\"](@/)?(app/|src/)?\.*(/)?(models/)",
                        f"from '{CORRECT_IMPORT}",
                        line
                    )
                    
                    issues.append({
                        'line_number': line_num,
                        'line': line.strip(),
                        'corrected': corrected.strip(),
                        'type': pattern['description'],
                    })
                    
                    # Atualizar estatísticas
                    desc = pattern['description']
                    stats['issues_by_type'][desc] = stats['issues_by_type'].get(desc, 0) + 1
                    stats['total_issues'] += 1
                    
    except Exception as e:
        print(f"{Colors.RED}Erro ao ler {file_path}: {e}{Colors.RESET}")
        
    return issues

def scan_directory(directory: str = '.'):
    """Escanear diretório recursivamente"""
    for root, dirs, files in os.walk(directory):
        # Filtrar diretórios
        dirs[:] = [d for d in dirs if not should_ignore(os.path.join(root, d))]
        
        for file in files:
            if any(file.endswith(ext) for ext in EXTENSIONS):
                file_path = os.path.join(root, file)
                stats['total_files'] += 1
                
                # Mostrar progresso
                if stats['total_files'] % 10 == 0:
                    print(f"\r{Colors.BLUE}   Escaneados: {stats['total_files']} arquivos...{Colors.RESET}", end='')
                
                issues = scan_file(file_path)
                if issues:
                    stats['problematic_files'].append({
                        'path': file_path,
                        'issues': issues,
                    })

def show_summary():
    """Exibir resumo"""
    print('\n')
    print(f"{Colors.BLUE}╔{'═' * 50}╗{Colors.RESET}")
    print(f"{Colors.BLUE}║{'📊 RESUMO'.center(50)}║{Colors.RESET}")
    print(f"{Colors.BLUE}╚{'═' * 50}╝{Colors.RESET}\n")
    
    print(f"{Colors.CYAN}📁 Total de arquivos escaneados: {Colors.GREEN}{stats['total_files']}{Colors.RESET}")
    print(f"{Colors.CYAN}📄 Arquivos com problemas: {Colors.RED}{len(stats['problematic_files'])}{Colors.RESET}")
    print(f"{Colors.CYAN}⚠️  Total de issues: {Colors.RED}{stats['total_issues']}{Colors.RESET}\n")
    
    if stats['issues_by_type']:
        print(f"{Colors.YELLOW}📋 Tipos de problemas encontrados:{Colors.RESET}\n")
        for issue_type, count in stats['issues_by_type'].items():
            print(f"  {Colors.RED}•{Colors.RESET} {issue_type}: {Colors.YELLOW}{count}{Colors.RESET} ocorrências")
        print()

def show_file_details(file_data: Dict):
    """Exibir detalhes de um arquivo"""
    print(f"{Colors.YELLOW}{'━' * 50}{Colors.RESET}")
    print(f"{Colors.CYAN}📄 Arquivo: {Colors.MAGENTA}{file_data['path']}{Colors.RESET}")
    print(f"{Colors.YELLOW}{'━' * 50}{Colors.RESET}\n")
    
    for issue in file_data['issues']:
        print(f"{Colors.RED}  ❌ Linha {issue['line_number']}:{Colors.RESET}")
        print(f"{Colors.BLUE}     {issue['line']}{Colors.RESET}")
        print(f"{Colors.GREEN}     ✅ Sugestão: {issue['corrected']}{Colors.RESET}\n")

def show_detailed_issues():
    """Exibir todos os arquivos problemáticos"""
    if not stats['problematic_files']:
        return
    
    print(f"{Colors.BLUE}╔{'═' * 50}╗{Colors.RESET}")
    print(f"{Colors.BLUE}║{'📝 ARQUIVOS PROBLEMÁTICOS'.center(50)}║{Colors.RESET}")
    print(f"{Colors.BLUE}╚{'═' * 50}╝{Colors.RESET}\n")
    
    for file_data in stats['problematic_files']:
        show_file_details(file_data)

def generate_report() -> str:
    """Gerar relatório em arquivo"""
    timestamp = datetime.now().strftime('%Y%m%d_%H%M%S')
    report_file = f'import-audit-report-{timestamp}.txt'
    
    with open(report_file, 'w', encoding='utf-8') as f:
        f.write('╔══════════════════════════════════════════════════╗\n')
        f.write('║     RELATÓRIO DE AUDITORIA DE IMPORTS           ║\n')
        f.write('╚══════════════════════════════════════════════════╝\n\n')
        f.write(f"Data: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}\n")
        f.write(f"Projeto: {os.path.basename(os.getcwd())}\n\n")
        f.write('════════════════════════════════════════════════════\n')
        f.write('RESUMO\n')
        f.write('════════════════════════════════════════════════════\n\n')
        f.write(f"Total de arquivos escaneados: {stats['total_files']}\n")
        f.write(f"Arquivos com problemas: {len(stats['problematic_files'])}\n")
        f.write(f"Total de issues encontrados: {stats['total_issues']}\n\n")
        
        if stats['problematic_files']:
            f.write('════════════════════════════════════════════════════\n')
            f.write('TIPOS DE PROBLEMAS\n')
            f.write('════════════════════════════════════════════════════\n\n')
            
            for issue_type, count in stats['issues_by_type'].items():
                f.write(f"  • {issue_type}: {count} ocorrências\n")
            
            f.write('\n════════════════════════════════════════════════════\n')
            f.write('ARQUIVOS PROBLEMÁTICOS\n')
            f.write('════════════════════════════════════════════════════\n\n')
            
            for file_data in stats['problematic_files']:
                f.write('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\n')
                f.write(f"📄 {file_data['path']}\n")
                f.write('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\n\n')
                
                for issue in file_data['issues']:
                    f.write(f"  ❌ Linha {issue['line_number']}:\n")
                    f.write(f"     {issue['line']}\n")
                    f.write(f"     ✅ Sugestão: {issue['corrected']}\n\n")
            
            f.write('════════════════════════════════════════════════════\n')
            f.write('COMANDO PARA CORRIGIR\n')
            f.write('════════════════════════════════════════════════════\n\n')
            f.write('Execute o script de migração para corrigir automaticamente:\n')
            f.write('  ./migrate-models.sh\n')
            f.write('  OU\n')
            f.write('  node migrate-models.js\n\n')
    
    return report_file

def show_final_result(report_file: str):
    """Exibir resultado final"""
    print()
    print(f"{Colors.BLUE}╔{'═' * 50}╗{Colors.RESET}")
    print(f"{Colors.BLUE}║{'🎯 RESULTADO FINAL'.center(50)}║{Colors.RESET}")
    print(f"{Colors.BLUE}╚{'═' * 50}╝{Colors.RESET}\n")
    
    if not stats['problematic_files']:
        print(f"{Colors.GREEN}✅ PARABÉNS! Todos os imports estão corretos!{Colors.RESET}")
        print(f"{Colors.GREEN}   Todos os arquivos usam: @/lib/db/models/{Colors.RESET}\n")
    else:
        print(f"{Colors.RED}❌ ATENÇÃO! Foram encontrados problemas nos imports{Colors.RESET}\n")
        print(f"{Colors.YELLOW}📄 Relatório detalhado salvo em:{Colors.RESET}")
        print(f"   {Colors.CYAN}{report_file}{Colors.RESET}\n")
        print(f"{Colors.YELLOW}🔧 Para corrigir automaticamente, execute:{Colors.RESET}")
        print(f"   {Colors.GREEN}./migrate-models.sh{Colors.RESET}")
        print(f"   {Colors.GREEN}OU{Colors.RESET}")
        print(f"   {Colors.GREEN}node migrate-models.js{Colors.RESET}\n")

def main():
    """Função principal"""
    print(f"{Colors.BLUE}╔{'═' * 50}╗{Colors.RESET}")
    print(f"{Colors.BLUE}║{'🔍 AUDITORIA DE IMPORTS - CRM SYSTEM'.center(50)}║{Colors.RESET}")
    print(f"{Colors.BLUE}╚{'═' * 50}╝{Colors.RESET}\n")
    
    # Verificar se está na raiz
    if not os.path.exists('package.json'):
        print(f"{Colors.RED}❌ Execute este script na raiz do projeto Next.js{Colors.RESET}")
        sys.exit(1)
    
    # Escanear projeto
    print(f"{Colors.YELLOW}🔍 Escaneando projeto...{Colors.RESET}\n")
    scan_directory()
    print(f"\n{Colors.GREEN}   ✅ Escaneamento concluído: {stats['total_files']} arquivos{Colors.RESET}\n")
    
    # Exibir resumo
    show_summary()
    
    # Gerar relatório
    report_file = generate_report()
    
    # Menu interativo
    if stats['problematic_files']:
        print(f"{Colors.CYAN}{'═' * 50}{Colors.RESET}")
        response = input(f"{Colors.YELLOW}Deseja ver os detalhes de cada arquivo? (s/n): {Colors.RESET}")
        if response.lower() == 's':
            show_detailed_issues()
    
    # Resultado final
    show_final_result(report_file)
    
    # Exit code
    sys.exit(1 if stats['problematic_files'] else 0)

if __name__ == '__main__':
    try:
        main()
    except KeyboardInterrupt:
        print(f"\n{Colors.YELLOW}⚠️  Auditoria cancelada pelo usuário{Colors.RESET}")
        sys.exit(1)
    except Exception as e:
        print(f"{Colors.RED}❌ Erro fatal: {e}{Colors.RESET}")
        sys.exit(1)

File: audit-imports.js
#!/usr/bin/env node

const fs = require('fs');
const path = require('path');

// Cores
const colors = {
  reset: '\x1b[0m',
  red: '\x1b[31m',
  green: '\x1b[32m',
  yellow: '\x1b[33m',
  blue: '\x1b[34m',
  magenta: '\x1b[35m',
  cyan: '\x1b[36m',
  dim: '\x1b[2m',
};

// Configuração
const config = {
  correctImport: '@/lib/db/models/',
  excludeDirs: ['node_modules', '.next', 'dist', 'build', '.git', 'coverage', '.turbo'],
  extensions: ['.ts', '.tsx'],
};

// Estatísticas
const stats = {
  totalFiles: 0,
  problematicFiles: [],
  issuesByType: {},
  totalIssues: 0,
  startTime: Date.now(),
  directoriesScanned: 0,
};

// Padrões de import incorretos
const wrongPatterns = [
  {
    regex: /from\s+['"]@\/models\//,
    description: "Import usando @/models/",
    priority: 'high',
  },
  {
    regex: /from\s+['"]@\/app\/models\//,
    description: "Import usando @/app/models/",
    priority: 'high',
  },
  {
    regex: /from\s+['"]@\/src\/models\//,
    description: "Import usando @/src/models/",
    priority: 'high',
  },
  {
    regex: /from\s+['"]\.\/models\//,
    description: "Import relativo ./models/",
    priority: 'medium',
  },
  {
    regex: /from\s+['"]\.\.\/models\//,
    description: "Import relativo ../models/",
    priority: 'medium',
  },
  {
    regex: /from\s+['"]\.\.\/\.\.\/models\//,
    description: "Import relativo ../../models/",
    priority: 'low',
  },
  {
    regex: /from\s+['"]\.\.\/\.\.\/\.\.\/models\//,
    description: "Import relativo ../../../models/",
    priority: 'low',
  },
];

// Funções de log
const log = {
  header: () => {
    console.log(`\n${colors.blue}╔${'═'.repeat(60)}╗${colors.reset}`);
    console.log(`${colors.blue}║${colors.cyan}  🔍 AUDITORIA DE IMPORTS - CRM SYSTEM  ${colors.blue}║${colors.reset}`);
    console.log(`${colors.blue}╚${'═'.repeat(60)}╝${colors.reset}\n`);
  },
  section: (title) => {
    console.log(`\n${colors.blue}╔${'═'.repeat(60)}╗${colors.reset}`);
    console.log(`${colors.blue}║${colors.cyan}${title.padStart(35).padEnd(60)}${colors.blue}║${colors.reset}`);
    console.log(`${colors.blue}╚${'═'.repeat(60)}╝${colors.reset}\n`);
  },
  progress: (current, dir) => {
    process.stdout.write(
      `\r${colors.cyan}📂 Escaneando: ${colors.dim}${dir.substring(0, 40).padEnd(40)}${colors.reset} ` +
      `${colors.yellow}(${current} arquivos)${colors.reset}   `
    );
  },
  error: (msg) => console.log(`${colors.red}❌ ${msg}${colors.reset}`),
  success: (msg) => console.log(`${colors.green}✅ ${msg}${colors.reset}`),
  warning: (msg) => console.log(`${colors.yellow}⚠️  ${msg}${colors.reset}`),
  info: (msg) => console.log(`${colors.blue}ℹ️  ${msg}${colors.reset}`),
};

// Verificar se deve ignorar diretório
function shouldIgnoreDir(dirPath) {
  const parts = dirPath.split(path.sep);
  
  // Verificar cada parte do caminho
  for (const part of parts) {
    // Ignorar diretórios da lista
    if (config.excludeDirs.includes(part)) {
      return true;
    }
    // Ignorar diretórios que começam com ponto (exceto .)
    if (part.startsWith('.') && part !== '.') {
      return true;
    }
    // Ignorar backups
    if (part.startsWith('backup_')) {
      return true;
    }
  }
  
  return false;
}

// Verificar se é arquivo válido
function isValidFile(fileName) {
  // Verificar extensão
  if (!config.extensions.some(ext => fileName.endsWith(ext))) {
    return false;
  }
  
  // Ignorar arquivos de teste e definição
  if (fileName.endsWith('.test.ts') || 
      fileName.endsWith('.test.tsx') || 
      fileName.endsWith('.spec.ts') || 
      fileName.endsWith('.spec.tsx') ||
      fileName.endsWith('.d.ts')) {
    return false;
  }
  
  return true;
}

// Escanear arquivo
function scanFile(filePath) {
  try {
    const content = fs.readFileSync(filePath, 'utf8');
    const lines = content.split('\n');
    const issues = [];

    lines.forEach((line, index) => {
      // Verificar cada padrão
      wrongPatterns.forEach(pattern => {
        if (pattern.regex.test(line)) {
          const lineNumber = index + 1;
          
          // Sugerir correção
          let corrected = line
            .replace(/from\s+['"]@\/models\//g, `from '${config.correctImport}`)
            .replace(/from\s+['"]@\/app\/models\//g, `from '${config.correctImport}`)
            .replace(/from\s+['"]@\/src\/models\//g, `from '${config.correctImport}`)
            .replace(/from\s+['"]\.+\/models\//g, `from '${config.correctImport}`);

          issues.push({
            lineNumber,
            line: line.trim(),
            corrected: corrected.trim(),
            type: pattern.description,
            priority: pattern.priority,
          });

          // Atualizar estatísticas
          stats.issuesByType[pattern.description] = 
            (stats.issuesByType[pattern.description] || 0) + 1;
          stats.totalIssues++;
        }
      });
    });

    return issues;
    
  } catch (err) {
    // Ignorar erros silenciosamente
    return [];
  }
}

// Escanear diretório recursivamente
function scanDirectory(dir) {
  try {
    // Verificar se deve ignorar este diretório
    if (shouldIgnoreDir(dir)) {
      return;
    }

    stats.directoriesScanned++;
    
    const entries = fs.readdirSync(dir, { withFileTypes: true });

    for (const entry of entries) {
      const fullPath = path.join(dir, entry.name);

      if (entry.isDirectory()) {
        // Recursão para subdiretórios
        scanDirectory(fullPath);
      } else if (entry.isFile() && isValidFile(entry.name)) {
        stats.totalFiles++;

        // Mostrar progresso
        if (stats.totalFiles % 5 === 0) {
          log.progress(stats.totalFiles, dir);
        }

        const issues = scanFile(fullPath);
        if (issues.length > 0) {
          stats.problematicFiles.push({
            path: fullPath,
            issues,
          });
        }
      }
    }
  } catch (err) {
    // Ignorar erros de permissão
  }
}

// Exibir resumo
function showSummary() {
  const elapsed = ((Date.now() - stats.startTime) / 1000).toFixed(2);
  const filesPerSecond = (stats.totalFiles / parseFloat(elapsed)).toFixed(0);
  
  console.log('\n'); // Limpar linha de progresso
  log.section('📊 RESUMO');

  console.log(`${colors.cyan}⏱️  Tempo de execução: ${colors.green}${elapsed}s${colors.reset}`);
  console.log(`${colors.cyan}⚡ Velocidade: ${colors.green}${filesPerSecond} files/s${colors.reset}`);
  console.log(`${colors.cyan}📂 Diretórios escaneados: ${colors.green}${stats.directoriesScanned}${colors.reset}`);
  console.log(`${colors.cyan}📁 Total de arquivos encontrados: ${colors.green}${stats.totalFiles}${colors.reset}`);
  console.log(`${colors.cyan}📄 Arquivos com problemas: ${colors.red}${stats.problematicFiles.length}${colors.reset}`);
  console.log(`${colors.cyan}⚠️  Total de issues: ${colors.red}${stats.totalIssues}${colors.reset}\n`);

  if (Object.keys(stats.issuesByType).length > 0) {
    console.log(`${colors.yellow}📋 Tipos de problemas encontrados:${colors.reset}\n`);
    
    // Ordenar por quantidade
    const sorted = Object.entries(stats.issuesByType)
      .sort((a, b) => b[1] - a[1]);
    
    sorted.forEach(([type, count]) => {
      const priority = wrongPatterns.find(p => p.description === type)?.priority || 'low';
      const priorityIcon = priority === 'high' ? '🔴' : priority === 'medium' ? '🟡' : '🔵';
      console.log(`  ${priorityIcon} ${type}: ${colors.yellow}${count}${colors.reset} ocorrências`);
    });
    console.log('');
  }
}

// Exibir detalhes de arquivo
function showFileDetails(fileData) {
  console.log(`${colors.yellow}${'━'.repeat(60)}${colors.reset}`);
  console.log(`${colors.cyan}📄 Arquivo: ${colors.magenta}${fileData.path}${colors.reset}`);
  console.log(`${colors.dim}   ${fileData.issues.length} issue(s) encontrado(s)${colors.reset}`);
  console.log(`${colors.yellow}${'━'.repeat(60)}${colors.reset}\n`);

  // Agrupar por prioridade
  const highPriority = fileData.issues.filter(i => i.priority === 'high');
  const mediumPriority = fileData.issues.filter(i => i.priority === 'medium');
  const lowPriority = fileData.issues.filter(i => i.priority === 'low');

  const showIssues = (issues, label, color) => {
    if (issues.length === 0) return;
    console.log(`${color}  ${label} (${issues.length}):${colors.reset}\n`);
    issues.forEach(issue => {
      console.log(`${colors.red}    ❌ Linha ${issue.lineNumber}:${colors.reset}`);
      console.log(`${colors.blue}       ${issue.line}${colors.reset}`);
      console.log(`${colors.green}       ✅ ${issue.corrected}${colors.reset}\n`);
    });
  };

  showIssues(highPriority, '🔴 ALTA PRIORIDADE', colors.red);
  showIssues(mediumPriority, '🟡 MÉDIA PRIORIDADE', colors.yellow);
  showIssues(lowPriority, '🔵 BAIXA PRIORIDADE', colors.blue);
}

// Exibir todos os arquivos problemáticos
function showDetailedIssues() {
  if (stats.problematicFiles.length === 0) return;

  log.section('📝 ARQUIVOS PROBLEMÁTICOS');

  // Ordenar por número de issues (decrescente)
  const sorted = [...stats.problematicFiles].sort((a, b) => b.issues.length - a.issues.length);

  sorted.forEach(fileData => {
    showFileDetails(fileData);
  });
}

// Gerar relatório em arquivo
function generateReport() {
  const timestamp = new Date().toISOString().replace(/[:.]/g, '-').split('T').join('_').substring(0, 19);
  const reportFile = `import-audit-report-${timestamp}.txt`;
  const elapsed = ((Date.now() - stats.startTime) / 1000).toFixed(2);
  const filesPerSecond = (stats.totalFiles / parseFloat(elapsed)).toFixed(0);

  let report = '';
  report += '╔══════════════════════════════════════════════════════════╗\n';
  report += '║        RELATÓRIO DE AUDITORIA DE IMPORTS                ║\n';
  report += '╚══════════════════════════════════════════════════════════╝\n\n';
  report += `Data: ${new Date().toLocaleString('pt-BR')}\n`;
  report += `Projeto: ${path.basename(process.cwd())}\n`;
  report += `Tempo de execução: ${elapsed}s\n`;
  report += `Velocidade: ${filesPerSecond} files/s\n\n`;
  report += '════════════════════════════════════════════════════════════\n';
  report += 'RESUMO\n';
  report += '════════════════════════════════════════════════════════════\n\n';
  report += `Diretórios escaneados: ${stats.directoriesScanned}\n`;
  report += `Total de arquivos encontrados: ${stats.totalFiles}\n`;
  report += `Arquivos com problemas: ${stats.problematicFiles.length}\n`;
  report += `Total de issues encontrados: ${stats.totalIssues}\n\n`;

  if (stats.problematicFiles.length > 0) {
    report += '════════════════════════════════════════════════════════════\n';
    report += 'TIPOS DE PROBLEMAS\n';
    report += '════════════════════════════════════════════════════════════\n\n';

    Object.entries(stats.issuesByType)
      .sort((a, b) => b[1] - a[1])
      .forEach(([type, count]) => {
        const priority = wrongPatterns.find(p => p.description === type)?.priority || 'low';
        report += `  [${priority.toUpperCase()}] ${type}: ${count} ocorrências\n`;
      });

    report += '\n════════════════════════════════════════════════════════════\n';
    report += 'ARQUIVOS PROBLEMÁTICOS (ordenados por quantidade de issues)\n';
    report += '════════════════════════════════════════════════════════════\n\n';

    const sorted = [...stats.problematicFiles].sort((a, b) => b.issues.length - a.issues.length);

    sorted.forEach(fileData => {
      report += '━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\n';
      report += `📄 ${fileData.path} (${fileData.issues.length} issues)\n`;
      report += '━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\n\n';

      fileData.issues.forEach(issue => {
        report += `  [${issue.priority.toUpperCase()}] Linha ${issue.lineNumber}:\n`;
        report += `     ❌ ${issue.line}\n`;
        report += `     ✅ ${issue.corrected}\n\n`;
      });
    });

    report += '════════════════════════════════════════════════════════════\n';
    report += 'COMANDO PARA CORRIGIR\n';
    report += '════════════════════════════════════════════════════════════\n\n';
    report += 'Execute o script de migração para corrigir automaticamente:\n';
    report += '  ./migrate-models.sh\n';
    report += '  OU\n';
    report += '  node migrate-models.js\n\n';
  }

  fs.writeFileSync(reportFile, report);
  return reportFile;
}

// Exibir resultado final
function showFinalResult(reportFile) {
  log.section('🎯 RESULTADO FINAL');

  if (stats.problematicFiles.length === 0) {
    log.success('PARABÉNS! Todos os imports estão corretos!');
    console.log(`${colors.green}   Todos os arquivos usam: @/lib/db/models/${colors.reset}\n`);
  } else {
    log.error('ATENÇÃO! Foram encontrados problemas nos imports\n');
    console.log(`${colors.yellow}📄 Relatório detalhado salvo em:${colors.reset}`);
    console.log(`   ${colors.cyan}${reportFile}${colors.reset}\n`);
    console.log(`${colors.yellow}🔧 Para corrigir automaticamente, execute:${colors.reset}`);
    console.log(`   ${colors.green}node migrate-models.js${colors.reset}\n`);
    
    console.log(`${colors.yellow}📋 Top 10 arquivos com mais problemas:${colors.reset}\n`);
    const sorted = [...stats.problematicFiles]
      .sort((a, b) => b.issues.length - a.issues.length)
      .slice(0, 10);
    
    sorted.forEach((file, index) => {
      const priorityIcon = file.issues.some(i => i.priority === 'high') ? '🔴' : 
                          file.issues.some(i => i.priority === 'medium') ? '🟡' : '🔵';
      console.log(`   ${priorityIcon} ${colors.red}${index + 1}.${colors.reset} ${file.path} ${colors.dim}(${file.issues.length} issues)${colors.reset}`);
    });
    console.log('');
  }
}

// Menu interativo
async function showMenu() {
  if (stats.problematicFiles.length === 0) return;

  const readline = require('readline').createInterface({
    input: process.stdin,
    output: process.stdout,
  });

  return new Promise(resolve => {
    console.log(`${colors.cyan}${'═'.repeat(60)}${colors.reset}`);
    readline.question(
      `${colors.yellow}Deseja ver os detalhes de cada arquivo? (s/n): ${colors.reset}`,
      answer => {
        readline.close();
        if (answer.toLowerCase() === 's') {
          showDetailedIssues();
        }
        resolve();
      }
    );
  });
}

// Listar estrutura de diretórios (debug)
function listDirectories(dir, level = 0) {
  if (level > 3) return; // Limitar profundidade
  
  try {
    const entries = fs.readdirSync(dir, { withFileTypes: true });
    
    entries.forEach(entry => {
      if (entry.isDirectory()) {
        const fullPath = path.join(dir, entry.name);
        if (!shouldIgnoreDir(fullPath)) {
          console.log(`${'  '.repeat(level)}📁 ${entry.name}`);
          listDirectories(fullPath, level + 1);
        }
      }
    });
  } catch (err) {
    // Ignorar erros
  }
}

// Função principal
async function main() {
  log.header();

  // Verificar se está na raiz
  if (!fs.existsSync('package.json')) {
    log.error('Execute este script na raiz do projeto Next.js');
    process.exit(1);
  }

  // Mostrar estrutura (debug)
  log.info('📂 Estrutura do projeto:');
  console.log('');
  listDirectories('.', 0);
  console.log('');

  // Escanear projeto
  log.info('🔍 Iniciando escaneamento recursivo...\n');
  scanDirectory('.');

  console.log('\n'); // Limpar linha de progresso

  if (stats.totalFiles === 0) {
    log.warning('Nenhum arquivo TypeScript encontrado!');
    log.info('Verifique se você está na raiz do projeto Next.js');
    process.exit(0);
  }

  // Exibir resumo
  showSummary();

  // Gerar relatório
  const reportFile = generateReport();
  log.success(`Relatório salvo em: ${colors.cyan}${reportFile}${colors.reset}`);

  // Menu interativo
  await showMenu();

  // Resultado final
  showFinalResult(reportFile);

  // Exit code
  process.exit(stats.problematicFiles.length > 0 ? 1 : 0);
}

// Executar
main().catch(err => {
  console.error(`${colors.red}Erro fatal:${colors.reset}`, err);
  process.exit(1);
});